wallet/wallet/hd_wallet.go

```diff
@@ -411,6 +413,75 @@ func (w *HDWallet) deriveKeyNodeV2(nextIndex uint32) (*slip10.Node, error) {
 	return keyNode, nil
 }
 
+func (w *HDWallet) removeTaintedKeyFromRestrictedKeys(taintedKeyPair HDKeyPair) {
+	allKeysAreTainted := w.areAllKeysTainted()
+
+	for hostname, permissions := range w.permissions {
+		if !permissions.PublicKeys.Enabled() {
+			continue
+		}
+
+		if !permissions.PublicKeys.HasRestrictedKeys() {
+			// If all the keys in the wallet are tainted, we revoke the
+			// permission.
+			if allKeysAreTainted {
+				permissions.PublicKeys = NoPublicKeysPermission()
+				w.permissions[hostname] = permissions
+			}
+			continue
+		}
+
+		restrictedKeys := permissions.PublicKeys.RestrictedKeys
+
+		// Look for the tainted key.
+		taintedKeyIdx := -1
+		for i, restrictedKey := range restrictedKeys {
+			if restrictedKey == taintedKeyPair.PublicKey() {
+				taintedKeyIdx = i
+				break
+			}
+		}
+
+		// No tainted key was found, next.
+		if taintedKeyIdx == -1 {
+			continue
+		}
+
+		lastItemIdx := len(restrictedKeys) - 1
+
+		// If lastItemIdx is 0, it means we have a single restricted key, and it
+		// is tainted. Removing it will make the slice empty.
+		// The user had a clear intent to restrict the access to this single
+		// key, we should void the access to public keys and let the third-party
+		// application request new permissions.
+		// This seems to be the least surprising behavior.
+		if lastItemIdx == 0 {
+			permissions.PublicKeys = NoPublicKeysPermission()
+		} else {
+			// We remove the key from the slice.
+			if taintedKeyIdx < lastItemIdx {
+				copy(restrictedKeys[taintedKeyIdx:], restrictedKeys[taintedKeyIdx+1:])
+			}
+			restrictedKeys[lastItemIdx] = ""
+			restrictedKeys = restrictedKeys[:lastItemIdx]
+			permissions.PublicKeys.RestrictedKeys = restrictedKeys
+		}
+
+		w.permissions[hostname] = permissions
+	}
+}
+
+func (w *HDWallet) areAllKeysTainted() bool {
+	allKeysAreTainted := true
+	for _, keyPair := range w.keyRing.ListKeyPairs() {
+		if !keyPair.IsTainted() {
```

Comment:
maybe:
```
if allKeysAreTainted&=!keyPair.IsTainted();allKeysAreTainted {
      break
}

```

Reply:
I dont' think it's more readable

Reply:
However, I found a better way to write this:
```golang
	for _, keyPair := range w.keyRing.ListKeyPairs() {
		if !keyPair.IsTainted() {
			return false
		}
	}
	return true
```

Reply:
Thanks form make me reflect on it